src/pkg/systemartifact/manager.go

```diff
@@ -0,0 +1,231 @@
+	RegisterCleanupCriteria(vendor string, artifactType string, criteria CleanupCriteria)
+
+	// GetCleanupCriteria returns a clean-up criteria for a specific vendor and artifact type combination.
+	// if no clean-up criteria is found then the default clean-up criteria is returned
+	GetCleanupCriteria(vendor string, artifactType string) CleanupCriteria
+
+	// Cleanup cleans up the system artifacts (tracking records as well as blobs) based on the
+	// artifact records selected by the CleanupCriteria registered for each vendor type.
+	// Returns the total number of records deleted, the reclaimed size and any error (if encountered)
+	Cleanup(ctx context.Context) (int64, int64, error)
+}
+
+type systemArtifactManager struct {
+	regCli          registry.Client
+	dao             dao.DAO
+	cleanupCriteria CleanupCriteria
+}
+
+func NewManager() Manager {
+	sysArtifactMgr := &systemArtifactManager{
+		regCli:          registry.Cli,
+		dao:             dao.NewSystemArtifactDao(),
+		cleanupCriteria: DefaultCleanupCriteria,
+	}
+	return sysArtifactMgr
+}
+
+func (mgr *systemArtifactManager) Create(ctx context.Context, artifactRecord *model.SystemArtifact, reader io.Reader) (int64, error) {
+
+	id, err := mgr.dao.Create(ctx, artifactRecord)
+
+	if err != nil {
+		return int64(0), err
+	}
+
+	repoName := mgr.getRepositoryName(artifactRecord.Vendor, artifactRecord.Repository)
+	err = mgr.regCli.PushBlob(repoName, artifactRecord.Digest, artifactRecord.Size, reader)
+
+	if err != nil {
+
+		mgr.dao.Delete(ctx, artifactRecord.Vendor, artifactRecord.Repository, artifactRecord.Digest)
+		return int64(0), err
+	}
+	return id, nil
+}
+
+func (mgr *systemArtifactManager) Read(ctx context.Context, vendor string, repository string, digest string) (io.ReadCloser, error) {
+	sa, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return nil, err
+	}
+	repoName := mgr.getRepositoryName(vendor, repository)
+	_, readCloser, err := mgr.regCli.PullBlob(repoName, sa.Digest)
+	if err != nil {
+		return nil, err
+	}
+	return readCloser, nil
+}
+
+func (mgr *systemArtifactManager) Delete(ctx context.Context, vendor string, repository string, digest string) error {
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	err := mgr.regCli.DeleteBlob(repoName, digest)
+
+	if err != nil {
+		return err
+	}
+
+	err = mgr.dao.Delete(ctx, vendor, repository, digest)
+
+	return err
+}
+
+func (mgr *systemArtifactManager) Exists(ctx context.Context, vendor string, repository string, digest string) (bool, error) {
+	_, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return false, err
+	}
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	exist, err := mgr.regCli.BlobExist(repoName, digest)
+
+	if err != nil {
+		return false, err
+	}
+
+	return exist, nil
+}
+
+func (mgr *systemArtifactManager) GetStorageSize(ctx context.Context) (int64, error) {
+	listQuery := q.Query{}
+	artifacts, err := mgr.dao.List(ctx, &listQuery)
```

Comment:
if there are a lot of artifacts, here maybe exist potential performance issue, how about use SUM SQL to aggregate the result?

Reply:
+1